core/src/main/java/org/apache/iceberg/rest/RESTTableOperations.java

```diff
@@ -0,0 +1,212 @@
+    } else {
+      this.current = current;
+    }
+  }
+
+  @Override
+  public TableMetadata current() {
+    return current;
+  }
+
+  @Override
+  public TableMetadata refresh() {
+    return updateCurrentMetadata(client.get(path, LoadTableResponse.class, ErrorHandlers.tableErrorHandler()));
+  }
+
+  @Override
+  public void commit(TableMetadata base, TableMetadata metadata) {
+    UpdateTableRequest.Builder requestBuilder;
+    List<MetadataUpdate> baseChanges;
+    Consumer<ErrorResponse> errorHandler;
+    switch (updateType) {
+      case CREATE:
+        Preconditions.checkState(base == null, "Invalid base metadata for create transaction, expected null: %s", base);
+        requestBuilder = UpdateTableRequest.builderForCreate();
+        baseChanges = createChanges;
+        errorHandler = ErrorHandlers.tableErrorHandler(); // throws NoSuchTableException
+        break;
+
+      case REPLACE:
+        Preconditions.checkState(base != null, "Invalid base metadata: null");
+        // use the original replace base metadata because the transaction will refresh
+        requestBuilder = UpdateTableRequest.builderForReplace(replaceBase);
+        baseChanges = createChanges;
+        errorHandler = ErrorHandlers.tableCommitHandler();
+        break;
+
+      case SIMPLE:
+        Preconditions.checkState(base != null, "Invalid base metadata: null");
+        requestBuilder = UpdateTableRequest.builderFor(base);
+        baseChanges = ImmutableList.of();
+        errorHandler = ErrorHandlers.tableCommitHandler();
+        break;
+
+      default:
+        throw new UnsupportedOperationException(String.format("Update type %s is not supported", updateType));
+    }
+
+    baseChanges.forEach(requestBuilder::update);
+    metadata.changes().forEach(requestBuilder::update);
+    UpdateTableRequest request = requestBuilder.build();
+
+    // the error handler will throw necessary exceptions like CommitFailedException and UnknownCommitStateException
+    // TODO: ensure that the HTTP client lib passes HTTP client errors to the error handler
+    LoadTableResponse response = client.post(path, request, LoadTableResponse.class, errorHandler);
+
+    // all future commits should be simple commits
+    this.updateType = UpdateType.SIMPLE;
+
+    updateCurrentMetadata(response);
+  }
+
+  @Override
+  public FileIO io() {
+    return io;
+  }
+
+  private TableMetadata updateCurrentMetadata(LoadTableResponse response) {
+    // LoadTableResponse is used to deserialize the response, but config is not allowed by the REST spec so it can be
+    // safely ignored. there is no requirement to update config on refresh or commit.
+    if (current == null || !Objects.equals(current.metadataFileLocation(), response.metadataLocation())) {
+      this.current = response.tableMetadata();
+    }
+
+    return current;
+  }
+
+  private static String metadataFileLocation(TableMetadata metadata, String filename) {
+    String metadataLocation = metadata.properties()
```

Comment:
nit: maybe rename to customMetadataLocation, otherwise the code below reads a bit like "if metadatalocation is null use metadata.location".

Reply:
`metadata.location()` is the table's location, not the metadata location.